### PR TITLE
refactor(components): share logic between refinement and menu components

### PR DESCRIPTION
**Pull request by francoischalifour**
> This is a big PR, mainly due to the types, snapshots and previously missing tests. I can try to split it in several PRs to make the review easier.

This refactors the widgets depending on the previous `RefinementList` component:

- `RefinementList`
- `HierarchicalMenu`
- `Menu`
- `NumericMenu`
- `RatingMenu`

## Impact

It will be easier to implement "Show More" and "Search For Facet Values" behaviors in their corresponding widgets.

They are also now typed with TypeScript.

## Dependencies

We now have a main component called `List` that handles:

- SearchBox
- Facets display
- No results display
- Show more

This is the component that all other list widgets use.

`HierarchicalMenu` and `SharedRefinementList` directely uses this `List` component. `SharedRefinementList` provides the primitives to build a refinement list experience. `Menu`, `NumericMenu` and `RatingMenu` use `SharedRefinementList`.

```
- List
  - HierarchicalMenu
  - SharedRefinementList
    - RefinementList
    - Menu
    - NumericMenu
    - RatingMenu
```

### CODE REVIEW

**Comment by eunjae-lee** [comment]
If I understand correctly, this is a different approach from the previous compositions with enhancers. Any reason you changed the implementation?

**Comment by francoischalifour** [comment]
"Enhancers" was a proposal at the connector level. This PR changes the architecture of how the logic is shared among all the Refinements/Menu components.

We had a discussion with AlexS two years ago about how difficult it was to maintain the RefinementList component because of the way it was designed. This is another take that makes it easier to add features for React InstantSearch parity:

- Search For Facet Values
- Show More

Again, this is only at the component level.

It also makes the components easier to test. They are now their own components, reusing the `SharedRefinementList`. I removed some dead code about a inexistent `depth` CSS class. Finally, they use TypeScript.

I plan to write tests for the `List` component, because nothing was tested so far.

File: tsconfig.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "noImplicitReturns": true,
     "noFallthroughCasesInSwitch": true,
     "allowSyntheticDefaultImports": true,
-    "skipLibCheck": true
+    "skipLibCheck": true,
+    "resolveJsonModule": true
```
[the remark this answers is not in the document]
Comment (francoischalifour) [reply tsconfig.json]: I use a JSON fixture here: https://github.com/algolia/instantsearch.js/blob/0ce05add4c5fac734292ca6d17c8e899c81ed955/src/components/HierarchicalMenu/__tests__/HierarchicalMenu-test.tsx#L6

I can convert it to JavaScript if needed.

**Comment by francoischalifour** [comment]
I'm closing this PR in the hope to open several ones better split later.

**Pull request closed without merging**